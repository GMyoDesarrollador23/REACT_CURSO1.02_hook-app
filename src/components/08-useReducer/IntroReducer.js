const inicialState = [
   {
      id: 1,
      todo: "Comprar pan",
      done: false,
   },
];

const newTodo = {
   id: 2,
   todo: "Comprar leche",
   done: false,
};

const agregarElTodoAction = {
   type: "agregar",
   payload: newTodo,
};

const todoReducer = (state = inicialState, action) => {
   if (action?.type === "agregar") {
      return [...state, action.payload];
   }

   return state;
};

let todos = todoReducer();

todos = todoReducer(todos, agregarElTodoAction);

console.log(todos);
