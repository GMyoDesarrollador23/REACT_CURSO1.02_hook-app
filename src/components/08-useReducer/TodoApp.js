import React, { useReducer, useEffect } from "react";

import { todoReducer } from "./todoreducer";
import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";

import "./style.css";

const init = () => {
   return JSON.parse(localStorage.getItem("todos")) || [];
};

export const TodoApp = () => {
   const [todos, dispatch] = useReducer(
      todoReducer,
      [],
      init
   ); 

   useEffect(() => {
      localStorage.setItem("todos", JSON.stringify(todos));
   }, [todos]);

   const handleDelate = (todoId) => {
      const action = {
         type: "delete",
         payload: todoId,
      };

      dispatch(action);
   };

   const handleToggle = (todoId) => {
      dispatch({
         type: "toggle",
         payload: todoId,
      });
   };

   const handleAddTodo = (newTodo) => {
      dispatch({
         type: "add",
         payload: newTodo,
      });
   };

   return (
      <div>
         <h1>TodoApp ({todos.length})</h1>
         <hr />

         <div className="row">
            <div className="col-7">
               todos
               <TodoList
                  todos={todos}
                  handleDelate={handleDelate}
                  handleToggle={handleToggle}
               />
            </div>

            <div className="col-5">
               <TodoAdd handleAddTodo={handleAddTodo} />
            </div>
         </div>
      </div>
   );
};
