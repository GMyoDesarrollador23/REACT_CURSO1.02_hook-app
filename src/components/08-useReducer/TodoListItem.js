import React from "react";

export const TodoListItem = ({
   todo,
   index,
   handleDelate,
   handleToggle,
}) => {
   return (
      <li key={todo.id} className="list-group-item">
         <p
            className={`${todo.done && "complite"}`}
            onClick={() => {
               handleToggle(todo.id);
            }}
         >
            {index + 1}.{todo.desc}
         </p>
 
         <button
            value={todo.id}
            onClick={() => {
               handleDelate(todo.id);
            }}
            className="btn btn-outline-danger"
         >
            Borrar
         </button>
      </li>
   );
};
