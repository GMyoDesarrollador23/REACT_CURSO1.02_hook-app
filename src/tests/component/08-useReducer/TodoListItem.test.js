import React from "react";
import { shallow } from "enzyme";
import "@testing-library/jest-dom";

import { TodoListItem } from "../../../components/08-useReducer/TodoListItem";
import { demoTodos } from "../../fixtures/demoTodos";

describe("Pruebas sobre TodoListItem", () => {
   const handleDelate = jest.fn();
   const handleToggle = jest.fn();

   const wrapper = shallow(
      <TodoListItem
         todo={demoTodos[0]}
         index={0}
         handleDelate={handleDelate}
         handleToggle={handleToggle}
      />
   );
   //---------------------------------------------------------
   test("debe de mostrarse correctamete", () => {
      expect(wrapper).toMatchSnapshot();
   });

   //---------------------------------------------------------
   test("debe de llamar la funcion de borrar", () => {
      wrapper.find("button").simulate("click");

      expect(handleDelate).toHaveBeenCalledWith(
         demoTodos[0].id
      );
   });
   //---------------------------------------------------------

   test("debe llamar la funcion toggle", () => {
      wrapper.find("p").simulate("click");

      expect(handleToggle).toHaveBeenCalledWith(
         demoTodos[0].id
      );
   });

   //---------------------------------------------------------
   test("debe de mostrar el texto correntamente", () => {
      const p = wrapper.find("p");
      
      expect(p.text().trim()).toBe(
         `1.${demoTodos[0].desc}`
      );
   });

   //---------------------------------------------------------
   test("debe tener la clase complete si el todo.done esta en true", () => {
      const todo = demoTodos[0];
      todo.done = true;

      const wrapper = shallow(<TodoListItem todo={todo} />);
      // console.log(wrapper.html());
      const p = wrapper.find("p");
      expect(p.hasClass("complite")).toBe(true);
   });

   //---------------------------------------------------------
});
