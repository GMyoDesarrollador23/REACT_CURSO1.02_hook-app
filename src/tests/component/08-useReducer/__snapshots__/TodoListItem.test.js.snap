// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre TodoListItem debe de mostrarse correctamete 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
    .
    aprender Recat
  </p>
  <button
    className="btn btn-outline-danger"
    onClick={[Function]}
    value={1}
  >
    Borrar
  </button>
</li>
`;
